Update torchrun and TorchElastic to take optional `local_addr` param to allow skip local IP lookup if specified

cz-37 opened the pull request:
Summary:
Update dynamic renderzvous nodes to use rendezvous hostname if provided.
For PR: https://github.com/pytorch/pytorch/issues/85300

Before:
For dynamic renderzvous, it always grab the `fqdn` from socket for each node even if user specified the address.
For example,
https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py#L248-L256
```
return _NodeDesc(socket.getfqdn(), os.getpid(), local_id)
```

Now:
If user specifies the hostname, each node will respect the given hostname.
For example, `socket.getfqdn(<hostname>) `

Test Plan: Unit tests.

Differential Revision: D41204028

d4l3k commented on torch/distributed/elastic/rendezvous/utils.py:
```diff
@@ -54,6 +54,21 @@ def _parse_rendezvous_config(config_str: str) -> Dict[str, str]:
     return config
 
 
+def _try_parse_host_from_endpoint(endpoint: str) -> Optional[str]:
```
can we add some unit tests for this parsing logic?

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1231,13 +1235,18 @@ def create_handler(
+        hostname = ''
+        param_hostname = _try_parse_host_from_endpoint(params.endpoint)
+        if param_hostname:
+            hostname = param_hostname
```
nit:
```
hostname = param_hostname if param_hostname else ''
```
or
```
hostname = param_hostname or ''
```

d4l3k commented on torch/distributed/elastic/rendezvous/dynamic_rendezvous.py:
```diff
@@ -1231,13 +1235,18 @@ def create_handler(
             _get_timeout(params, "close"),
         )
+        hostname = ''
+        param_hostname = _try_parse_host_from_endpoint(params.endpoint)
```
This endpoint could be a different host right?

like if you rendezvous with endpoint == "rank0" on rank 1 machine we wouldn't want to grab rank0 as the current address? am I misunderstanding this?

d4l3k reviewed (approved):
LGTM see internal comments

cz-37 committed:
Update torchrun and TorchElastic to take optional `local_addr` param to allow skip local IP lookup if specified (#88922)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/88922

Update dynamic renderzvous nodes to use rendezvous hostname if provided.

Before:
For dynamic renderzvous, it always grab the `fqdn` from socket for each node even if user specified the address.
For example,
https://github.com/pytorch/pytorch/blob/master/torch/distributed/elastic/rendezvous/dynamic_rendezvous.py#L248-L256
```
return _NodeDesc(socket.getfqdn(), os.getpid(), local_id)
```

Now:
If user specifies the hostname, each node will respect the given hostname.
For example, `socket.getfqdn(<hostname>) `

Test Plan:
```
% python -m torch.distributed.run  --help
usage: run.py [-h] [--nnodes NNODES] [--nproc_per_node NPROC_PER_NODE] [--rdzv_backend RDZV_BACKEND] [--rdzv_endpoint RDZV_ENDPOINT] [--rdzv_id RDZV_ID] [--rdzv_conf RDZV_CONF] [--standalone] [--max_restarts MAX_RESTARTS] [--monitor_interval MONITOR_INTERVAL]
              [--start_method {spawn,fork,forkserver}] [--role ROLE] [-m] [--no_python] [--run_path] [--log_dir LOG_DIR] [-r

% python -m torch.distributed.run  --local_addr 192.168.0.84 --rdzv_backend 'c10d'  training_script.py
NOTE: Redirects are currently not supported in Windows or MacOs.
Hello World

% python -m torch.distributed.run   --rdzv_backend 'c10d'  training_script.py
NOTE: Redirects are currently not supported in Windows or MacOs.
Hello World

```

Unit tests with the 2 added tests.

Reviewed By: d4l3k

Differential Revision: D41204028

fbshipit-source-id: 6dbf07c10f19c44e691e7dcaa4cb3df62c735a98
